### PR TITLE
Integrate Std_MeasureDistance into unified measurement facility

### LINKED ISSUE #13712: Measurement: Support of individual picking points for distance
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Currently when selecting two lines, surfaces or features in the new unified measurement tool, the points to calculate a distance are selected automatically by the tool, it seems to choose the closest ones.
It should be possible to select individual picking points that are on the selected features, e.g. select the point on a line or surface from which the distance should be calculated.

### Full version info

```shell
OS: Windows 10 build 19045
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.37063 (Git)
Build type: Release
Branch: main
Hash: 680792030fd664c94829e2c1b78b21788e8d9879
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: German/Germany (de_DE)
Installed mods: 
  * BIM 2021.12.0
  * CurvedShapes 1.0.9
  * Curves 0.6.35
  * fasteners 0.5.20
  * OpenTheme 2024.4.20
  * OpticsWorkbench 1.0.17
  * sheetmetal 0.4.13

```

### Subproject(s) affected?

Core

### Anything else?

https://github.com/FreeCAD/FreeCAD/pull/9750#issuecomment-2042250708

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hlorus**
Currently the Std_MeasureDistance is still present as it's functionality wasn't completly covered by UMF's MeasureDistance. This was because the UMF version would only measure between actual geometry elements rather than  the locations where the user is clicking.

This PR adds a new measure type "MeasureDistanceDetached" to UMF which allows to measure a distance between two locations on the surface of objects. 
Also the old command "Std_MeasureDistance" along with the feature "App/MeasureDistance.h" and it's viewprovider "Gui/MeasureDistance.h" get removed here.

Closes #13712

### CODE REVIEW

**Comment by adrianinsaval** [comment]
If I'm not mistaken the old command added a document object in the tree which was saved with the document, what happens when opening old files that have these objects? should be looking into some sort of migration code? @wwmayer do you know how we can migrate old objects to a new object type? is it worth it in this case?

**Comment by hlorus** [comment]
> > what happens when opening old files that have these objects?
> 
> A message will be printed that the object couldn't be restored.
> 
> > do you know how we can migrate old objects to a new object type?
> 
> A migration is not easily possible because there is no mechanism that the document could ask to create an alternative object if the MeasureDistance cannot be instantiated. A workaround could be to implement a special Type object for MeasureDistance that instantiates the corresponding feature type of the Measure module. This feature then must handle the properties of MeasureDistance.

Could we also keep the old object around and migrate it in the onDocumentRestored method similar to how it's documented here: https://wiki.freecad.org/Scripted_objects_migration ?
The old document could then be removed sometime in the future...

**Comment by wwmayer** [comment]
> Could we also keep the old object around and migrate it in the onDocumentRestored method similar to how it's documented here: https://wiki.freecad.org/Scripted_objects_migration ?

No, this will cause serious problems:
onDocumentRestored would have to instantiate the new object, migrate the properties and then destroy itself. At the time when this method is used the document iterates over its objects and the iterator will become invalid. Another problem is that the pointer the document uses to reference this object will become dangling.

The class MeasureDistance is part of the core system and it would directly import an extension module that is considered bad practise. Lower level modules must not directly depend anyhow on higher level modules as this breaks the design.

**Comment by adrianinsaval** [comment]
the question is now is it worth trying to support those objects? is anybody using them?

**Comment by wwmayer** [comment]
I guess it's used rather rarely. But nevertheless with this trick the object can be restored. Add the following code snippets to Mod/Measure/MeasureDistance.h/cpp:
```
class MeasureDistanceType: public Base::BaseClass
{
public:
    static Base::Type getClassTypeId();
    Base::Type getTypeId() const override;
    static void init();
    static void* create();

private:
    static Base::Type classTypeId;
};
```
and
```
Base::Type MeasureDistanceType::getClassTypeId()
{
    return Base::Type::badType();
}

Base::Type MeasureDistanceType::getTypeId() const
{
    return Base::Type::badType();
}

void MeasureDistanceType::init()
{
    initSubclass(MeasureDistanceType::classTypeId,
                 "App::MeasureDistance",
                 "App::DocumentObject",
                 &(MeasureDistanceType::create));
}

void* MeasureDistanceType::create()
{
    return new MeasureDistance();
}

Base::Type MeasureDistanceType::classTypeId = Base::Type::badType();
```
In order to read in the properties P1 and P2 MeasureDistance must override handleChangedPropertyName:
```
void MeasureDistance::handleChangedPropertyName(Base::XMLReader &reader,
                                                const char * TypeName,
                                                const char *PropName)
{
    if (strcmp(PropName, "P1") == 0 && strcmp(TypeName, "App::PropertyVector") == 0) {
        Position1.Restore(reader);
    }
    else if (strcmp(PropName, "P2") == 0 && strcmp(TypeName, "App::PropertyVector") == 0) {
        Position2.Restore(reader);
    }
}
```

**Comment by hlorus** [comment]
> @hlorus are you adding the snippets above?

Yeah sure, soonish

**Comment by hlorus** [comment]
> I guess it's used rather rarely. But nevertheless with this trick the object can be restored. Add the following code snippets to Mod/Measure/MeasureDistance.h/cpp:
> 
> ```
> class MeasureDistanceType: public Base::BaseClass
> {
> public:
>     static Base::Type getClassTypeId();
>     Base::Type getTypeId() const override;
>     static void init();
>     static void* create();
> 
> private:
>     static Base::Type classTypeId;
> };
> ```
> 
> and
> 
> ```
> Base::Type MeasureDistanceType::getClassTypeId()
> {
>     return Base::Type::badType();
> }
> 
> Base::Type MeasureDistanceType::getTypeId() const
> {
>     return Base::Type::badType();
> }
> 
> void MeasureDistanceType::init()
> {
>     initSubclass(MeasureDistanceType::classTypeId,
>                  "App::MeasureDistance",
>                  "App::DocumentObject",
>                  &(MeasureDistanceType::create));
> }
> 
> void* MeasureDistanceType::create()
> {
>     return new MeasureDistance();
> }
> 
> Base::Type MeasureDistanceType::classTypeId = Base::Type::badType();
> ```
> 
> In order to read in the properties P1 and P2 MeasureDistance must override handleChangedPropertyName:
> 
> ```
> void MeasureDistance::handleChangedPropertyName(Base::XMLReader &reader,
>                                                 const char * TypeName,
>                                                 const char *PropName)
> {
>     if (strcmp(PropName, "P1") == 0 && strcmp(TypeName, "App::PropertyVector") == 0) {
>         Position1.Restore(reader);
>     }
>     else if (strcmp(PropName, "P2") == 0 && strcmp(TypeName, "App::PropertyVector") == 0) {
>         Position2.Restore(reader);
>     }
> }
> ```

Alrigth, i've added the snippets in a new commit and it seems to work well. For some reason the newly created measurements are not visible tho.

**Comment by wwmayer** [comment]
> For some reason the newly created measurements are not visible tho.

I have observed this too but couldn't spot the code block where the view provider fails to show the label. However, if you set the Element links, save the project and reload it the label is shown.

But anyway, it at least shows the distance in the tree view and this is still better than nothing.

**Comment by hlorus** [comment]
I found the reason why it wasn't displaying and added a commit.

**Comment by hlorus** [comment]
> I can't select two points on the same face. Is it normal? The old tool allowed for that.

That's indeed a limitation as we're using the global selection state and the face is deselected on the second click.

**Comment by FEA-eng** [comment]
> That's indeed a limitation as we're using the global selection state and the face is deselected on the second click.

Ok, I see. Do you plan to fix it later on? Should I create an issue? There are the new selection modes but it's either face, edge or vertex selection so it may not help here.

**Comment by maxwxyz** [comment]
Is the name "Distance Free" ok?

**Comment by FEA-eng** [comment]
> Is the name "Distance Free" ok?

Might be better to name this "Distance Detached" or "Distance Arbitrary".

**Comment by maxwxyz** [comment]
I think the type in the UI and Tree should just say "Distance" as it's measuring the distance. Maybe a checkbox could be implemented instead of choosing between features of the model or free picking points. That would also suit further extensions, e.g. snapping to centerlines of holes or shafts or using axes (https://github.com/FreeCAD/FreeCAD/issues/13708) or project dimension (x,y,z) as the old tool (https://github.com/FreeCAD/FreeCAD/issues/13741). 
If we get for each new picking point or mode for measuring a distance a new "Distance ..." entry in the tool's dropdown and as tree label, I think this is bad UI.
@pierreporte you've created a lot of issues regarding this tool, what's your opinion?

@FreeCAD/design-working-group maybe others can weigh in on this.

**Comment by pierreporte** [comment]
@maxwxyz As I understand, the choice is about letting users to measure distance between features, and between two arbitrary points on the geometry. I think that it’s a bit limited. Sometimes you want to measure the distance between an edge and an arbitrary point. The idea is to have something similar to Catia that lets you chose the type of the two objects to measure the distance between.

(I’m not really available this week so I may have missed something to take my comment with a grain of salt.)

**Comment by wwmayer** [comment]
> and between two arbitrary points on the geometry.

You can measure the distance between two arbitrary vertexes but not between two arbitrary points. The advantage is that the measurement is parametric but the disadvantage is that is lacks of flexibility.

A workaround is to use the Part_Primitives command and create two vertexes. Thanks to the 3D View button it allows to define the vertexes on arbitrary positions. This way it's possible to measure distances on meshes or point clouds.

**Comment by maxwxyz** [comment]
> @maxwxyz As I understand, the choice is about letting users to measure distance between features, and between two arbitrary points on the geometry. I think that it’s a bit limited.

No, what I meant was that as currently implemented you'll end up with 3+ items in that dropdown as measurement type although you want to measure a distance. It'll be: Distance, Distance Free, Distance Snapping, Distance ...
Better would be just the  Type distance and specify for both points how they are picked: Based on features or a free picking point. Ideally you can choose for each picking point individually, e.g. measure the dimension of a feature (e.g. an edge) and a free picking point. Or measure between a centerline of a cylindrical surface (snapping) and a face.
This solution would allow the mix of multiple picking points for distance.

**Comment by FEA-eng** [comment]
To be honest, measurement between arbitrary (non-vertex) points is not so common. But there are some situations when it’s useful (quick approximate measure, no vertices in the desired locations, measuring on imported images and so on). So I wouldn’t mind if it was a separate mode but Max is right that adding even more modes can clutter the tool so checkboxes seem to be a good solution. Even if they limit the choice to either 2 vertices or 2 detached points as it was the case with the old tools as well.

**Comment by pierreporte** [comment]
@maxwxyz We actually agree. My previous comment wasn’t clear enough.

**Comment by hlorus** [comment]
> > That's indeed a limitation as we're using the global selection state and the face is deselected on the second click.
> 
> Ok, I see. Do you plan to fix it later on? Should I create an issue? There are new selection modes but it's either face, edge or vertex selection so it may not help here.

Yeah these modes won't help, we would likely need a selection mode which doesn't deselect elements on second click but rather just adds the new picking location to the list. However i'm not sure if the selection system easily allows for that.

**Comment by hlorus** [comment]
> I think the type in the UI and Tree should just say "Distance" as it's measuring the distance. Maybe a checkbox could be implemented instead of choosing between features of the model or free picking points. That would also suit further extensions, e.g. snapping to centerlines of holes or shafts or using axes (#13708) or project dimension (x,y,z) as the old tool (#13741). If we get for each new picking point or mode for measuring a distance a new "Distance ..." entry in the tool's dropdown and as tree label, I think this is bad UI. @pierreporte you've created a lot of issues regarding this tool, what's your opinion?
> 
> @FreeCAD/design-working-group maybe others can weigh in on this.

I agree that it should ideally be an additional setting, however this might be a bit confusing if there's only one mode which actually supports it. I think people would expect that measure types such as angle, radius and position also work with arbitrary picking points.

**Comment by maxwxyz** [comment]
Maybe the possible picking modes can be dependent on the measurement type.
CATIA does it like that: You select for both selection points the method.
catiameasure2
This is the possible selection for each point:
grafik

@hlorus demo:
https://youtu.be/bBd6L1-oaaA?si=O_xmXQbJ5247Fzw-&t=59

**Comment by chennes** [comment]
This needs a rebase to fix the conflicts.

File: src/3rdParty/OndselSolver
Comment (adrianinsaval) [inline src/3rdParty/OndselSolver:1]: please remove this from this commit

**Comment by chennes** [comment]
There is a conflict that needs to be resolved. @adrianinsaval is this good to go in from your perspective?

**Pull request merged**